File: TelemetryDashboard.WPF/Services/TelemetryService.cs
using System;
using System.IO.Ports;
using System.Text;
using System.Threading;
using System.Threading.Tasks;
using TelemetryDashboard.Models;

namespace TelemetryDashboard.Services
{
    /// <summary>
    /// Handles serial port communication and telemetry data decoding
    /// </summary>
    public class TelemetryService : IDisposable
    {
        private SerialPort? _serialPort;
        private CancellationTokenSource? _cancellationTokenSource;
        private Task? _readTask;
        private bool _isRunning;

        public event EventHandler<TelemetryData>? TelemetryReceived;
        public event EventHandler<string>? ErrorOccurred;
        public event EventHandler<string>? StatusChanged;

        public string PortName { get; private set; }
        public int BaudRate { get; private set; }
        public bool IsConnected => _serialPort?.IsOpen ?? false;
        public bool UseAsciiMode { get; set; } = true;

        public TelemetryService(string portName = "COM3", int baudRate = 115200)
        {
            PortName = portName;
            BaudRate = baudRate;
        }

        /// <summary>
        /// Get available serial ports on the system
        /// </summary>
        public static string[] GetAvailablePorts()
        {
            return SerialPort.GetPortNames();
        }

        /// <summary>
        /// Connect to the serial port
        /// </summary>
        public bool Connect()
        {
            try
            {
                _serialPort = new SerialPort
                {
                    PortName = PortName,
                    BaudRate = BaudRate,
                    DataBits = 8,
                    Parity = Parity.None,
                    StopBits = StopBits.One,
                    Handshake = Handshake.None,
                    ReadTimeout = 1000,
                    WriteTimeout = 1000
                };

                _serialPort.Open();
                _serialPort.DiscardInBuffer();
                _serialPort.DiscardOutBuffer();

                StatusChanged?.Invoke(this, $"Connected to {PortName} at {BaudRate} baud");
                return true;
            }
            catch (Exception ex)
            {
                ErrorOccurred?.Invoke(this, $"Failed to connect: {ex.Message}");
                return false;
            }
        }

        /// <summary>
        /// Disconnect from the serial port
        /// </summary>
        public void Disconnect()
        {
            StopReading();

            if (_serialPort?.IsOpen == true)
            {
                _serialPort.Close();
                StatusChanged?.Invoke(this, "Disconnected from serial port");
            }

            _serialPort?.Dispose();
            _serialPort = null;
        }

        /// <summary>
        /// Start reading telemetry data in background
        /// </summary>
        public void StartReading()
        {
            if (_isRunning || _serialPort == null || !_serialPort.IsOpen)
                return;

            _isRunning = true;
            _cancellationTokenSource = new CancellationTokenSource();

            _readTask = Task.Run(() => ReadLoop(_cancellationTokenSource.Token));
            StatusChanged?.Invoke(this, "Started reading telemetry data");
        }

        /// <summary>
        /// Stop reading telemetry data
        /// </summary>
        public void StopReading()
        {
            if (!_isRunning)
                return;

            _isRunning = false;
            _cancellationTokenSource?.Cancel();

            try
            {
                _readTask?.Wait(TimeSpan.FromSeconds(2));
            }
            catch (AggregateException)
            {
                // Task was cancelled
            }

            _cancellationTokenSource?.Dispose();
            _cancellationTokenSource = null;
            _readTask = null;

            StatusChanged?.Invoke(this, "Stopped reading telemetry data");
        }

        /// <summary>
        /// Main reading loop
        /// </summary>
        private void ReadLoop(CancellationToken cancellationToken)
        {
            while (_isRunning && !cancellationToken.IsCancellationRequested)
            {
                try
                {
                    TelemetryData? telemetry = UseAsciiMode
                        ? ReadAsciiData()
                        : ReadBinaryData();

                    if (telemetry != null)
                    {
                        TelemetryReceived?.Invoke(this, telemetry);
                    }

                    Thread.Sleep(10); // Small delay to prevent CPU spinning
                }
                catch (TimeoutException)
                {
                    // Normal timeout, continue
                }
                catch (Exception ex)
                {
                    if (_isRunning) // Only report errors if still running
                    {
                        ErrorOccurred?.Invoke(this, $"Read error: {ex.Message}");
                        Thread.Sleep(1000); // Wait before retrying
                    }
                }
            }
        }

        /// <summary>
        /// Read ASCII formatted telemetry data
        /// Expected format: "TEMP:25.5,PRESS:101.3,ALT:1234.5,VEL:100.2,ACCEL:9.8"
        /// </summary>
        private TelemetryData? ReadAsciiData()
        {
            if (_serialPort == null || !_serialPort.IsOpen)
                return null;

            try
            {
                string line = _serialPort.ReadLine().Trim();
                if (string.IsNullOrEmpty(line))
                    return null;

                var data = new TelemetryData();
                var pairs = line.Split(',');

                foreach (var pair in pairs)
                {
                    var parts = pair.Split(':');
                    if (parts.Length != 2)
                        continue;

                    string key = parts[0].Trim();
                    if (!float.TryParse(parts[1].Trim(), out float value))
                        continue;

                    switch (key.ToUpper())
                    {
                        case "TEMP":
                            data.Temperature = value;
                            break;
                        case "PRESS":
                            data.Pressure = value;
                            break;
                        case "ALT":
                            data.Altitude = value;
                            break;
                        case "VEL":
                            data.Velocity = value;
                            break;
                        case "ACCEL":
                            data.Acceleration = value;
                            break;
                    }
                }

                return data;
            }
            catch (TimeoutException)
            {
                throw; // Propagate timeout
            }
            catch (Exception ex)
            {
                ErrorOccurred?.Invoke(this, $"ASCII parsing error: {ex.Message}");
                return null;
            }
        }

        /// <summary>
        /// Read binary telemetry data
        /// Packet format (24 bytes):
        /// - Temperature: float (4 bytes)
        /// - Pressure: float (4 bytes)
        /// - Altitude: float (4 bytes)
        /// - Velocity: float (4 bytes)
        /// - Acceleration: float (4 bytes)
        /// - Checksum: uint32 (4 bytes)
        /// Start marker: 0xFF 0xAA
        /// </summary>
        private TelemetryData? ReadBinaryData()
        {
            if (_serialPort == null || !_serialPort.IsOpen)
                return null;

            try
            {
                // Look for start marker (0xFF 0xAA)
                while (true)
                {
                    int byte1 = _serialPort.ReadByte();
                    if (byte1 != 0xFF)
                        continue;

                    int byte2 = _serialPort.ReadByte();
                    if (byte2 != 0xAA)
                        continue;

                    // Found start marker, read packet (24 bytes)
                    byte[] buffer = new byte[24];
                    int bytesRead = _serialPort.Read(buffer, 0, 24);

                    if (bytesRead == 24)
                    {
                        return DecodeBinaryPacket(buffer);
                    }

                    break;
                }

                return null;
            }
            catch (TimeoutException)
            {
                throw; // Propagate timeout
            }
            catch (Exception ex)
            {
                ErrorOccurred?.Invoke(this, $"Binary read error: {ex.Message}");
                return null;
            }
        }

        /// <summary>
        /// Decode binary telemetry packet
        /// </summary>
        private TelemetryData? DecodeBinaryPacket(byte[] buffer)
        {
            try
            {
                if (buffer.Length < 24)
                    return null;

                float temperature = BitConverter.ToSingle(buffer, 0);
                float pressure = BitConverter.ToSingle(buffer, 4);
                float altitude = BitConverter.ToSingle(buffer, 8);
                float velocity = BitConverter.ToSingle(buffer, 12);
                float acceleration = BitConverter.ToSingle(buffer, 16);
                uint checksum = BitConverter.ToUInt32(buffer, 20);

                // Verify checksum (sum of first 20 bytes)
                uint calculatedChecksum = 0;
                for (int i = 0; i < 20; i++)
                {
                    calculatedChecksum += buffer[i];
                }

                if (calculatedChecksum != checksum)
                {
                    ErrorOccurred?.Invoke(this, "Checksum mismatch in binary packet");
                    return null;
                }

                return new TelemetryData(temperature, pressure, altitude, velocity, acceleration);
            }
            catch (Exception ex)
            {
                ErrorOccurred?.Invoke(this, $"Binary decode error: {ex.Message}");
                return null;
            }
        }

        /// <summary>
        /// Send command to device
        /// </summary>
        public bool SendCommand(string command)
        {
            if (_serialPort == null || !_serialPort.IsOpen)
                return false;

            try
            {
                _serialPort.WriteLine(command);
                return true;
            }
            catch (Exception ex)
            {
                ErrorOccurred?.Invoke(this, $"Send error: {ex.Message}");
                return false;
            }
        }

        public void Dispose()
        {
            Disconnect();
        }
    }
}

File: telemetry_service.py
"""
Telemetry Service - Serial Port Reader
Reads and decodes telemetry data from a serial port.
"""
import serial
import json
import struct
import time
from datetime import datetime
from typing import Dict, Any, Optional
import threading


class TelemetryData:
    """Represents a telemetry data packet"""

    def __init__(self, timestamp: str, temperature: float, pressure: float,
                 altitude: float, velocity: float, acceleration: float):
        self.timestamp = timestamp
        self.temperature = temperature
        self.pressure = pressure
        self.altitude = altitude
        self.velocity = velocity
        self.acceleration = acceleration

    def to_dict(self) -> Dict[str, Any]:
        """Convert to dictionary for JSON serialization"""
        return {
            'timestamp': self.timestamp,
            'temperature': self.temperature,
            'pressure': self.pressure,
            'altitude': self.altitude,
            'velocity': self.velocity,
            'acceleration': self.acceleration
        }


class TelemetryService:
    """Handles serial port communication and telemetry data decoding"""

    def __init__(self, port: str = 'COM3', baudrate: int = 115200):
        """
        Initialize telemetry service

        Args:
            port: Serial port name (e.g., 'COM3' on Windows, '/dev/ttyUSB0' on Linux)
            baudrate: Serial communication speed
        """
        self.port = port
        self.baudrate = baudrate
        self.serial_connection: Optional[serial.Serial] = None
        self.is_running = False
        self.data_callback = None
        self.thread: Optional[threading.Thread] = None

    def connect(self) -> bool:
        """
        Connect to the serial port

        Returns:
            True if connection successful, False otherwise
        """
        try:
            self.serial_connection = serial.Serial(
                port=self.port,
                baudrate=self.baudrate,
                timeout=1.0,
                bytesize=serial.EIGHTBITS,
                parity=serial.PARITY_NONE,
                stopbits=serial.STOPBITS_ONE
            )
            print(f"Connected to {self.port} at {self.baudrate} baud")
            return True
        except serial.SerialException as e:
            print(f"Failed to connect: {e}")
            return False

    def disconnect(self):
        """Disconnect from the serial port"""
        self.stop_reading()
        if self.serial_connection and self.serial_connection.is_open:
            self.serial_connection.close()
            print("Disconnected from serial port")

    def decode_telemetry_packet(self, data: bytes) -> Optional[TelemetryData]:
        """
        Decode binary telemetry packet

        Packet format (24 bytes):
        - Temperature: float (4 bytes)
        - Pressure: float (4 bytes)
        - Altitude: float (4 bytes)
        - Velocity: float (4 bytes)
        - Acceleration: float (4 bytes)
        - Checksum: uint32 (4 bytes)

        Args:
            data: Raw binary data

        Returns:
            TelemetryData object or None if decoding fails
        """
        try:
            if len(data) < 24:
                return None

            # Unpack binary data (little-endian format)
            values = struct.unpack('<ffffI', data[:24])
            temperature = values[0]
            pressure = values[1]
            altitude = values[2]
            velocity = values[3]
            acceleration = values[4]
            checksum = values[5]

            # Simple checksum verification (sum of first 20 bytes)
            calculated_checksum = sum(data[:20]) & 0xFFFFFFFF
            if calculated_checksum != checksum:
                print("Checksum mismatch!")
                return None

            timestamp = datetime.now().isoformat()

            return TelemetryData(
                timestamp=timestamp,
                temperature=temperature,
                pressure=pressure,
                altitude=altitude,
                velocity=velocity,
                acceleration=acceleration
            )
        except struct.error as e:
            print(f"Error decoding packet: {e}")
            return None

    def read_line_ascii(self) -> Optional[TelemetryData]:
        """
        Read ASCII formatted telemetry data (alternative to binary)

        Expected format: "TEMP:25.5,PRESS:101.3,ALT:1234.5,VEL:100.2,ACCEL:9.8"

        Returns:
            TelemetryData object or None if parsing fails
        """
        try:
            if not self.serial_connection or not self.serial_connection.is_open:
                return None

            line = self.serial_connection.readline().decode('utf-8').strip()
            if not line:
                return None

            # Parse key-value pairs
            data_dict = {}
            for pair in line.split(','):
                if ':' in pair:
                    key, value = pair.split(':')
                    data_dict[key.strip()] = float(value.strip())

            timestamp = datetime.now().isoformat()

            return TelemetryData(
                timestamp=timestamp,
                temperature=data_dict.get('TEMP', 0.0),
                pressure=data_dict.get('PRESS', 0.0),
                altitude=data_dict.get('ALT', 0.0),
                velocity=data_dict.get('VEL', 0.0),
                acceleration=data_dict.get('ACCEL', 0.0)
            )
        except Exception as e:
            print(f"Error reading ASCII data: {e}")
            return None

    def read_binary_packet(self) -> Optional[TelemetryData]:
        """
        Read binary telemetry packet from serial port

        Returns:
            TelemetryData object or None if reading fails
        """
        try:
            if not self.serial_connection or not self.serial_connection.is_open:
                return None

            # Look for start marker (0xFF 0xAA)
            while True:
                byte1 = self.serial_connection.read(1)
                if not byte1 or byte1[0] != 0xFF:
                    continue

                byte2 = self.serial_connection.read(1)
                if not byte2 or byte2[0] != 0xAA:
                    continue

                # Found start marker, read packet
                packet_data = self.serial_connection.read(24)
                if len(packet_data) == 24:
                    return self.decode_telemetry_packet(packet_data)
                break

            return None
        except Exception as e:
            print(f"Error reading binary packet: {e}")
            return None

    def start_reading(self, callback, use_ascii: bool = True):
        """
        Start reading telemetry data in a background thread

        Args:
            callback: Function to call with each TelemetryData object
            use_ascii: If True, read ASCII format; if False, read binary format
        """
        self.data_callback = callback
        self.is_running = True

        def read_loop():
            while self.is_running:
                try:
                    if use_ascii:
                        telemetry = self.read_line_ascii()
                    else:
                        telemetry = self.read_binary_packet()

                    if telemetry and self.data_callback:
                        self.data_callback(telemetry)

                    time.sleep(0.01)  # Small delay to prevent CPU spinning
                except Exception as e:
                    print(f"Error in read loop: {e}")
                    time.sleep(1)

        self.thread = threading.Thread(target=read_loop, daemon=True)
        self.thread.start()
        print("Started telemetry reading thread")

    def stop_reading(self):
        """Stop the background reading thread"""
        self.is_running = False
        if self.thread:
            self.thread.join(timeout=2.0)
            print("Stopped telemetry reading thread")


def main():
    """Example usage"""
    def on_telemetry_received(telemetry: TelemetryData):
        print(f"Received: {json.dumps(telemetry.to_dict(), indent=2)}")

    service = TelemetryService(port='COM3', baudrate=115200)

    if service.connect():
        service.start_reading(on_telemetry_received, use_ascii=True)

        try:
            # Keep running until Ctrl+C
            while True:
                time.sleep(1)
        except KeyboardInterrupt:
            print("\nShutting down...")
        finally:
            service.disconnect()


if __name__ == '__main__':
    main()
